reference/shaders-msl/comp/integer-dot-product.comp

```diff
@@ -0,0 +1,50 @@
+#pragma clang diagnostic ignored "-Wmissing-prototypes"
+
+#include <metal_stdlib>
+#include <simd/simd.h>
+
+using namespace metal;
+
+char reduce_add(char2 x) { return x.x + x.y; }
+char reduce_add(char3 x) { return x.x + x.y + x.z; }
+char reduce_add(char4 x) { return x.x + x.y + x.z + x.w; }
+uchar reduce_add(uchar2 x) { return x.x + x.y; }
+uchar reduce_add(uchar3 x) { return x.x + x.y + x.z; }
+uchar reduce_add(uchar4 x) { return x.x + x.y + x.z + x.w; }
+short reduce_add(short2 x) { return x.x + x.y; }
+short reduce_add(short3 x) { return x.x + x.y + x.z; }
+short reduce_add(short4 x) { return x.x + x.y + x.z + x.w; }
+ushort reduce_add(ushort2 x) { return x.x + x.y; }
+ushort reduce_add(ushort3 x) { return x.x + x.y + x.z; }
+ushort reduce_add(ushort4 x) { return x.x + x.y + x.z + x.w; }
```

Comment:
Can this really not be made into a template?

```
template <typename T, int count>
T reduce_add2(vector<T, count> v) { return v.x + v.y; }
```
etc. Maybe it can be specialized for count 2, 3, and 4, but probably not worth the hassle.